### ISSUE TITLE
kv: node down with closed timestamp regression

### ISSUE THREAD

**Issue by ajstorm** [cockroachdb/cockroach#121480]
Over the weekend, the `drt-large` cluster lost a node with:

```
I240331 15:40:23.115317 7909 1@kv/kvserver/replica_closedts.go:263 ⋮ [n14] 1067593  the server is terminating due to a fatal error (see the DEV channel for details)
F240331 15:40:23.116782 7909 kv/kvserver/replica_closedts.go:263 ⋮ [n14] 1067594  side-transport update saw closed timestamp regression on r15717: (lai=9, ts=1711899619.871648444,0) -> (lai=8, ts=1711899619.880149791,0)
```

A debug.zip exists on `drt-chaos-workload:1` (named `closed_timestamp_regression_debug.zip`), but as this cluster is on spot instances with local SSDs, if you need any additional information from the data on this node, I'd encourage you to act urgently, as the data could disappear at any time without notice.

Jira issue: CRDB-37283

Epic CRDB-37617

**Comment by kvoli** [comment]
Possibly hitting https://github.com/cockroachdb/cockroach/issues/70894 here. I will link the debug zip once uploaded to gdrive.

The regression is between the LAI.

**Comment by kvoli** [comment]
[Debug zip link](https://drive.google.com/file/d/1kq5jtM2JAi8qR5s7sz5t4LjXBh0ISb_s/view?usp=sharing) (internal).

**Comment by kvoli** [comment]
The lease transfers to `n14` (`n7` -> `n15` -> `n14`) approx. 2s earlier to it encountering a closed TS update where the LAI regressed.

```
I240331 15:40:21.847816 158258164 13@kv/kvserver/lease_queue.go:143 ⋮ [T1,Vsystem,n5,lease,s5,r15717/4:‹/Table/114/1/"`"/130{28…-73…}›] 356126 transferring lease to s14 usage=[batches/s=0.0 request_cpu/s=0µs raft_cpu/s=0µs write(keys)/s=0.0 write(bytes)/s=0 B read(keys)/s=0.0 read(bytes)/s=0 B]
F240331 15:40:23.116782 7909 kv/kvserver/replica_closedts.go:263 ⋮ [T1,n14] 1067594 side-transport update saw closed timestamp regression on r15717: (lai=9, ts=1711899619.871648444,0) -> (lai=8, ts=1711899619.880149791,0)
```

<details><summary>Full</summary>
<p>

```
I240331 15:33:44.582495 445 kv/kvserver/store_raft.go:699 ⋮ [T1,Vsystem,n5,s5,r15717/4:‹/Table/114/1/"`"/130{28…-73…}›,raft] 861278 raft ready handling: 0.68s [append=0.00s, apply=0.48s, non-blocking-sync=0.20s, other=0.00s], wrote [append-batch=51 B, apply=9 B (1)]; node might be overloaded
W240331 15:37:45.897510 453 2@kv/kvserver/replica_proposal.go:624 ⋮ [T1,Vsystem,n15,s15,r15717/1:‹/Table/114/1/"`"/130{28…-73…}›,raft] 234678 applied lease after ~1.84s replication lag, client traffic may have been delayed [lease=repl=(n15,s15):1 seq=44 start=1711854873.301391646,0 epo=331 pro=1711899464.053843963,0 prev=repl=(n15,s15):1 seq=43 start=1711854873.301391646,0 epo=303 pro=1711899224.555197515,0]
W240331 15:37:45.744589 156338462 kv/kvserver/liveness/liveness.go:752 ⋮ [T1,Vsystem,n15,s15,r15717/1:‹/Table/114/1/"`"/130{28…-73…}›] 2024763 slow heartbeat took 1.689608426s; err=<nil>
W240331 15:37:45.897510 453 2@kv/kvserver/replica_proposal.go:624 ⋮ [T1,Vsystem,n15,s15,r15717/1:‹/Table/114/1/"`"/130{28…-73…}›,raft] 2025037 applied lease after ~1.84s replication lag, client traffic may have been delayed [lease=repl=(n15,s15):1 seq=44 start=1711854873.301391646,0 epo=331 pro=1711899464.053843963,0 prev=repl=(n15,s15):1 seq=43 start=1711854873.301391646,0 epo=303 pro=1711899224.555197515,0]
W240331 15:40:10.787574 156513877 kv/kvserver/liveness/liveness.go:752 ⋮ [T1,Vsystem,n15,s15,r15717/1:‹/Table/114/1/"`"/130{28…-73…}›] 2046375 slow heartbeat took 4.731809114s; err=<nil>
I240331 15:40:17.215231 155919208 13@kv/kvserver/lease_queue.go:143 ⋮ [T1,Vsystem,n7,lease,s7,r15717/2:‹/Table/114/1/"`"/130{28…-73…}›] 376006 transferring lease to s15 usage=[batches/s=0.0 request_cpu/s=0µs raft_cpu/s=0µs write(keys)/s=0.0 write(bytes)/s=0 B read(keys)/s=0.0 read(bytes)/s=0 B]
W240331 15:40:18.230241 450 2@kv/kvserver/replica_proposal.go:624 ⋮ [T1,Vsystem,n15,s15,r15717/1:‹/Table/114/1/"`"/130{28…-73…}›,raft] 2047359 applied lease after ~1.01s replication lag, client traffic may have been delayed [lease=repl=(n15,s15):1 seq=46 start=1711899617.215333513,0 exp=1711899623.215274719,0 pro=1711899617.215274719,0 prev=repl=(n7,s7):2 seq=45 start=1711899612.637811165,0 epo=28 pro=1711899612.637811165,0]
W240331 15:40:18.230241 450 2@kv/kvserver/replica_proposal.go:624 ⋮ [T1,Vsystem,n15,s15,r15717/1:‹/Table/114/1/"`"/130{28…-73…}›,raft] 238199 applied lease after ~1.01s replication lag, client traffic may have been delayed [lease=repl=(n15,s15):1 seq=46 start=1711899617.215333513,0 exp=1711899623.215274719,0 pro=1711899617.215274719,0 prev=repl=(n7,s7):2 seq=45 start=1711899612.637811165,0 epo=28 pro=1711899612.637811165,0]
I240331 15:40:20.449417 446 kv/kvserver/store_raft.go:699 ⋮ [T1,Vsystem,n15,s15,r15717/1:‹/Table/114/1/"`"/130{28…-73…}›,raft] 2047666 raft ready handling: 0.80s [append=0.00s, apply=0.00s, non-blocking-sync=0.80s, other=0.00s], wrote [append-batch=51 B, ]; node might be overloaded
I240331 15:40:20.451451 495 kv/kvserver/replica_raft.go:1615 ⋮ [T1,Vsystem,n15,s15,r15717/1:‹/Table/114/1/"`"/130{28…-73…}›,raft] 2047771 slow non-blocking raft commit: commit-wait 498.207µs sem 192ns
I240331 15:40:21.847816 158258164 13@kv/kvserver/lease_queue.go:143 ⋮ [T1,Vsystem,n5,lease,s5,r15717/4:‹/Table/114/1/"`"/130{28…-73…}›] 356126 transferring lease to s14 usage=[batches/s=0.0 request_cpu/s=0µs raft_cpu/s=0µs write(keys)/s=0.0 write(bytes)/s=0 B read(keys)/s=0.0 read(bytes)/s=0 B]
F240331 15:40:23.116782 7909 kv/kvserver/replica_closedts.go:263 ⋮ [T1,n14] 1067594 side-transport update saw closed timestamp regression on r15717: (lai=9, ts=1711899619.871648444,0) -> (lai=8, ts=1711899619.880149791,0)
```

</p>
</details> 

<details><summary>Stack</summary>
<p>


```
F240331 15:40:23.116782 7909 kv/kvserver/replica_closedts.go:263 ⋮ [T1,n14] 1067594 side-transport update saw closed timestamp regression on r15717: (lai=9, ts=1711899619.871648444,0) -> (lai=8, ts=1711899619.880149791,0)
stack trace:
goroutine 7909 [running]:
github.com/cockroachdb/cockroach/pkg/util/allstacks.GetWithBuf({0x0?, 0xc126df73c0?, 0x4?})
	github.com/cockroachdb/cockroach/pkg/util/allstacks/allstacks.go:32 +0x74
github.com/cockroachdb/cockroach/pkg/util/allstacks.Get(...)
	github.com/cockroachdb/cockroach/pkg/util/allstacks/allstacks.go:19
github.com/cockroachdb/cockroach/pkg/util/log.(*loggerT).outputLogEntry(_, {{{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}}, ...})
	github.com/cockroachdb/cockroach/pkg/util/log/clog.go:282 +0xc6
github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal({0x7d041d0, 0xc0019f85a0}, 0x2, 0x4, 0x0, 0x0?, {0x66e7a03, 0x60}, {0xc126df7a00, 0x5, ...})
	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:109 +0x5c5
github.com/cockroachdb/cockroach/pkg/util/log.logfDepth(...)
	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:39
github.com/cockroachdb/cockroach/pkg/util/log.Fatalf(...)
	github.com/cockroachdb/cockroach/bazel-out/aarch64-opt/bin/pkg/util/log/log_channels_generated.go:848
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*sidetransportAccess).assertNoRegression(0x207?, {0x7d041d0, 0xc0019f85a0}, {{0x17c1e3a25d936abc, 0x0}, 0x9}, {{0x17c1e3a25e15231f, 0x0}, 0x8})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_closedts.go:263 +0x1cc
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*sidetransportAccess).forward(0xc0356985d0, {0x7d041d0, 0xc0019f85a0}, {0xc010314b68?, 0xffffffff?}, 0x8, 0x0)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_closedts.go:237 +0x11e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).ForwardSideTransportClosedTimestamp(...)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_closedts.go:138
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).ForwardSideTransportClosedTimestampForRange.func1(0xc010314bd0?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:133 +0x71
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).VisitStores.func1(0xc0019f85a0?, 0xc00c4b7500?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:150 +0x26
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*IntMap).Range(0xc1046f?, 0xc126df7c60)
	github.com/cockroachdb/cockroach/pkg/util/syncutil/int_map.go:385 +0xd6
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).VisitStores(0xc00b53e620?, 0x2b5?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:149 +0x4e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).ForwardSideTransportClosedTimestampForRange(0x5a75460?, {0x7d041d0, 0xc0019f85a0}, 0x7d041d0?, {0xc0019f85a0?, 0x46e600?}, 0x1?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:130 +0x85
github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/sidetransport.(*incomingStream).processUpdate(0xc00b464210, {0x7d041d0, 0xc0019f85a0}, 0xc25e167e60)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/sidetransport/receiver.go:253 +0x5bf
github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/sidetransport.(*incomingStream).Run.func1({0x7d041d0, 0xc0019f85a0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/sidetransport/receiver.go:329 +0x1a5
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x13a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 7908
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x3fe

```


</p>
</details> 

The debug zip ranges.json isn't as helpful here as it was captured a few hours after the closed TS regression.


From `n14` for `r15717`:

<details><summary>Details</summary>
<p>


```
[
  {
    "span": {
      "start_key": "/Table/114/1/\"`\"/13028/9/783",
      "end_key": "/Table/114/1/\"`\"/13073/10/1584"
    },
    "raft_state": {
      "replica_id": 3,
      "hard_state": {
        "term": 17,
        "vote": 1,
        "commit": 68
      },
      "lead": 1,
      "state": "StateFollower",
      "applied": 68,
      "progress": null
    },
    "state": {
      "state": {
        "raft_applied_index": 68,
        "lease_applied_index": 10,
        "desc": {
          "range_id": 15717,
          "start_key": "9nKJEmAAAfcy5JH3Aw8=",
          "end_key": "9nKJEmAAAfczEZL3BjA=",
          "internal_replicas": [
            {
              "node_id": 15,
              "store_id": 15,
              "replica_id": 1,
              "type": 0
            },
            {
              "node_id": 7,
              "store_id": 7,
              "replica_id": 2,
              "type": 0
            },
            {
              "node_id": 14,
              "store_id": 14,
              "replica_id": 3,
              "type": 0
            },
            {
              "node_id": 5,
              "store_id": 5,
              "replica_id": 4,
              "type": 0
            },
            {
              "node_id": 10,
              "store_id": 10,
              "replica_id": 5,
              "type": 0
            }
          ],
          "next_replica_id": 6,
          "generation": 553,
          "sticky_bit": {}
        },
        "lease": {
          "start": {
            "wall_time": 1711899629901676508
          },
          "replica": {
            "node_id": 15,
            "store_id": 15,
            "replica_id": 1,
            "type": 0
          },
          "proposed_ts": {
            "wall_time": 1711899630085767934
          },
          "epoch": 347,
          "sequence": 50,
          "acquisition_type": 2
        },
        "truncated_state": {
          "index": 10,
          "term": 5
        },
        "gc_threshold": {},
        "stats": {
          "contains_estimates": 0,
          "last_update_nanos": 1711722985669946118,
          "lock_age": 0,
          "gc_bytes_age": 0,
          "live_bytes": 272323524,
          "live_count": 453801,
          "key_bytes": 12651464,
          "key_count": 453801,
          "val_bytes": 259672060,
          "val_count": 453801,
          "intent_bytes": 0,
          "intent_count": 0,
          "lock_bytes": 0,
          "lock_count": 0,
          "range_key_count": 0,
          "range_key_bytes": 0,
          "range_val_count": 0,
          "range_val_bytes": 0,
          "sys_bytes": 474,
          "sys_count": 6,
          "abort_span_bytes": 0
        },
        "version": {
          "major": 23,
          "minor": 2,
          "patch": 0,
          "internal": 24
        },
        "raft_closed_timestamp": {
          "wall_time": 1711899626901814028
        },
        "raft_applied_index_term": 17,
        "gc_hint": {
          "latest_range_delete_timestamp": {},
          "gc_timestamp": {},
          "gc_timestamp_next": {}
        }
      },
      "last_index": 68,
      "raft_log_size": 1415,
      "range_max_bytes": 536870912,
      "active_closed_timestamp": {
        "wall_time": 1711908900025885848
      },
      "tenant_id": 1,
      "closed_timestamp_sidetransport_info": {
        "replica_closed": {
          "wall_time": 1711908900025885848
        },
        "replica_lai": 10,
        "central_closed": {
          "wall_time": 1711908900025885848
        },
        "central_lai": 10
      }
    },
    "source_node_id": 14,
    "source_store_id": 14,
    "lease_history": [
      {
        "start": {
          "wall_time": 1711899628673258634
        },
        "replica": {
          "node_id": 5,
          "store_id": 5,
          "replica_id": 4,
          "type": 0
        },
        "proposed_ts": {
          "wall_time": 1711899628673258634
        },
        "epoch": 53,
        "sequence": 49,
        "acquisition_type": 2
      },
      {
        "start": {
          "wall_time": 1711899629901676508
        },
        "expiration": {
          "wall_time": 1711899635901622381
        },
        "replica": {
          "node_id": 15,
          "store_id": 15,
          "replica_id": 1,
          "type": 0
        },
        "proposed_ts": {
          "wall_time": 1711899629901622381
        },
        "sequence": 50,
        "acquisition_type": 1
      },
      {
        "start": {
          "wall_time": 1711899629901676508
        },
        "replica": {
          "node_id": 15,
          "store_id": 15,
          "replica_id": 1,
          "type": 0
        },
        "proposed_ts": {
          "wall_time": 1711899630085767934
        },
        "epoch": 347,
        "sequence": 50,
        "acquisition_type": 2
      }
    ],
    "problems": {},
    "stats": {
      "writes_per_second": 0.0020311833623821956,
      "cpu_time_per_second": 893.600839831258
    },
    "lease_status": {
      "lease": {
        "start": {
          "wall_time": 1711899629901676508
        },
        "replica": {
          "node_id": 15,
          "store_id": 15,
          "replica_id": 1,
          "type": 0
        },
        "proposed_ts": {
          "wall_time": 1711899630085767934
        },
        "epoch": 347,
        "sequence": 50,
        "acquisition_type": 2
      },
      "now": {
        "wall_time": 1711908903006278394
      },
      "request_time": {
        "wall_time": 1711908903006278394
      },
      "state": 1,
      "liveness": {
        "node_id": 15,
        "epoch": 347,
        "expiration": {
          "wall_time": 1711908908055064068,
          "logical": 0
        }
      },
      "min_valid_observed_timestamp": {}
    },
    "quiescent": true,
    "top_k_locks_by_wait_queue_waiters": null,
    "locality": {
      "tiers": [
        {
          "key": "cloud",
          "value": "gce"
        },
        {
          "key": "region",
          "value": "northamerica-northeast1"
        },
        {
          "key": "zone",
          "value": "northamerica-northeast1-b"
        }
      ]
    },
    "lease_valid": true
  }
]
```

</p>
</details>

**Comment by kvoli** [comment]
From https://github.com/cockroachdb/cockroach/issues/122016#issuecomment-2045826880 the things we have found in common between the reproductions of this issue are:

- Lease is recently applied, <4s from regression detection
- Lease application is delayed by >1s of replication lag
- LAI regresses by 1, the TS increases

**Comment by pav-kv** [comment]
Is there commonality w.r.t. which node crashed? Is it always the new leaseholder, the old leaseholder, or some other node?

**Comment by kvoli** [comment]
There is not commonality, in 2/3 cases it is the new leaseholder. In the other case it is the [prior leaseholder](https://github.com/cockroachdb/cockroach/issues/122016#issuecomment-2045826880) which crashes. Although, we only know that it attempted to transfer the lease, there aren't any corresponding lease application logs on the receiver.

**Comment by dt** [comment]
also took down drt-ua1:4 today

**Comment by pav-kv** [comment]
@ajstorm @dt While we investigate this, can we enable `--vmodule=sender=4,receiver=4` on this cluster? If this happens again (given that it happened >= 4 times already), we should have more information in logs about the closed timestamp side-transport activity.

**Comment by pav-kv** [comment]
On `drt-ua1:4`, there is nothing happening with the range 20 min before it panicked.

```python
E240416 02:58:40.686206 1200056981 kv/kvclient/kvcoord/dist_sender_circuit_breaker.go:1003 ⋮ [T1,Vsystem,n4] 963207  r132799/3:(n2,s2) circuit breaker tripped: probe failed: [NotLeaseHolderError] ‹lease acquisition canceled because context canceled›; r132799: replica (n2,s2):3 not lease holder; lease holder unknown (stalled for 0s, erroring for 14.314s)
E240416 02:58:45.969310 1200066702 kv/kvclient/kvcoord/dist_sender_circuit_breaker.go:1003 ⋮ [T1,Vsystem,n4] 963286  r132799/2:(n3,s3) circuit breaker tripped: probe failed: [NotLeaseHolderError] ‹failed to manipulate liveness record: heartbeat failed on epoch increment›; r132799: replica (n3,s3):2 not lease holder; lease holder unknown (stalled for 3.981s, erroring for 4.282s)
F240416 03:20:59.333628 6891 kv/kvserver/replica_closedts.go:263 ⋮ [T1,n4] 978307  side-transport update saw closed timestamp regression on r132799: (lai=1022462, ts=1713237656.046687520,0) -> (lai=1022456, ts=1713237656.082527193,0)
```

There are a lot of "circuit breaker tripped" and "failed to manipulate liveness record" errors like the above.

**Comment by arulajmani** [comment]
Removing the GA-blocker label here given we saw this on a 23.2 cluster. This is still something we want to fix outside of blocker labels.

**Comment by nvanbenschoten** [comment]
Having now seen this once more on `drt-large`, I have a theory for what's going wrong.

In each of these cases, we see a combination of 1) disk slowness, 2) high replication lag, 3) lease instability, 4) liveness instability. I think these are all playing a role in this assertion failure.

In the latest failure, we saw the assertion:
```
drt-large-0015> F240430 18:12:12.094964 3807 kv/kvserver/replica_closedts.go:263 ⋮ [T1,Vsystem,n24] 1973086 side-transport update saw closed timestamp regression on r73332: (lai=4628420, ts=1714500726.759280263,0) -> (lai=4628419, ts=1714500728.593043438,0)
stack trace:
goroutine 3807 [running]:
...
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*sidetransportAccess).assertNoRegression(0x17c00000018?, {0x7cf5530, 0xc001cb7110}, {{0x17cb2153c1a2ce87, 0x0}, 0x469fc4}, {{0x17cb21542eefcfee, 0x0}, 0x469fc3})
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_closedts.go:263 +0x1cc
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*sidetransportAccess).forward(0xc187673350, {0x7cf5530, 0xc001cb7110}, {0x17c00000018?, 0x4?}, 0x469fc3, 0x1)
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_closedts.go:240 +0x15a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).BumpSideTransportClosed(0xc187672d88, {0x7cf5530, 0xc001cb7110}, {0x4?, 0x657ef99?}, {{0x17cb21542eefcfee, 0x0}, {0x17cb2155116f35ee, 0x0}})
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_closedts.go:109 +0x508
github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/sidetransport.(*Sender).publish(0xc00aafcb60, {0x7cf5530, 0xc001cb7110})
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/sidetransport/sender.go:381 +0x77a
github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/sidetransport.(*Sender).Run.func2({0x7cf5530, 0xc001cb7110})
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/sidetransport/sender.go:258 +0x193
```
Like we've seen in each of these cases, the closed timestamp regression was across exactly one lease-applied index and the closed timestamp regressed by more than a second

The stack trace of this assertion is interesting. In most of the cases, we've seen the regression get detected on the receiver-side of the closed timestamp transport. However, in this case and in https://github.com/cockroachdb/cockroach/issues/122016#issuecomment-2045892103, we actually see it on the sender side. That means that a leaseholder itself is detecting the regression when attempting to close a timestamp. That's kind of mindbending, as it means that a leaseholder has somehow learned of a higher lease-applied index than it has itself applied. This alone would not be an issue, except that we also check that the leaseholder has nothing in flight as part of `BumpSideTransportClosed` before getting to this assertion, so where did this `lai=4628420` come from?

Luckily, we have just enough information scattered across the logs to start to put together a lease history for this range. From the `applied lease after ~3.18s replication lag` and `transferring lease to` logs, we can construct this partial history:
```
repl=(n4,s59):1   seq=36 start=1714500688.686251546,0 epo=2 pro=1714500688.686251546,0
repl=(n24,s380):4 seq=37 start=1714500725.685074511,0 exp=1714500731.684506954,0 pro=1714500725.684506954,0
repl=(n4,s59):1   seq=38 ? (inferred)
repl=(n24,s380):4 seq=39 start=1714500729.844661546,0 exp=1714500735.844616045,0 pro=1714500729.844616045,0
repl=(n4,s59):1   seq=40 start=1714500735.844616045,1 epo=2 pro=1714500736.186413108,0
```

These come from this log on n24:
```
drt-large-0015> W240430 18:12:08.862871 2398 2@kv/kvserver/replica_proposal.go:624 ⋮ [T1,Vsystem,n24,s380,r73332/4:‹/Table/134/1/"`"/20{0/7…-6/6…}›,raft] 1972792 applied lease after ~3.18s replication lag, client traffic may have been delayed [lease=repl=(n24,s380):4 seq=37 start=1714500725.685074511,0 exp=1714500731.684506954,0 pro=1714500725.684506954,0 prev=repl=(n4,s59):1 seq=36 start=1714500688.686251546,0 epo=2 pro=1714500688.686251546,0]
```
and these logs on n4:
```
drt-large-0004> I240430 18:12:05.684459 6213562794 13@kv/kvserver/lease_queue.go:143 ⋮ [T1,Vsystem,n4,lease,s59,r73332/1:‹/Table/134/1/"`"/20{0/7…-6/6…}›] 737591 transferring lease to s380 usage=[batches/s=1.4 request_cpu/s=1ms raft_cpu/s=464µs write(keys)/s=3.3 write(bytes)/s=526 B read(keys)/s=7.7 read(bytes)/s=2.3 KiB]
drt-large-0004> I240430 18:12:09.844567 6213708992 13@kv/kvserver/lease_queue.go:143 ⋮ [T1,Vsystem,n4,lease,s59,r73332/1:‹/Table/134/1/"`"/20{0/7…-6/6…}›] 737676 transferring lease to s380 usage=[batches/s=0.0 request_cpu/s=0µs raft_cpu/s=0µs write(keys)/s=0.0 write(bytes)/s=0 B read(keys)/s=0.0 read(bytes)/s=0 B]
drt-large-0004> W240430 18:12:16.951549 1339 2@kv/kvserver/replica_proposal.go:637 ⋮ [T1,Vsystem,n4,s59,r73332/1:‹/Table/134/1/"`"/20{0/7…-6/6…}›,raft] 1082725 lease expired before epoch lease upgrade, client traffic may have been delayed [lease=repl=(n4,s59):1 seq=40 start=1714500735.844616045,1 epo=2 pro=1714500736.186413108,0 prev=repl=(n24,s380):4 seq=39 start=1714500729.844661546,0 exp=1714500735.844616045,0 pro=1714500729.844616045,0]
```
The key clues here are that:
1. the lease at sequence 39 has a start timestamp before the expiration of the lease at sequence 37
2. n4 transfers the lease to n24 at `18:12:05.684459` and at `18:12:09.844567`
3. n24 never transfers the lease to n4
4. we see a `lease expired before epoch lease upgrade` log for the transition from seq=39 to seq=40, but _not_ for the transition from seq=37 to seq=38. This implies that seq=37 was not an expiration-based lease at the time that it was replaced

We don't quite know details about the lease at sequence 38, but we can infer that it was owned by n4 and that it had a start time before the expiration of the lease at sequence 39.

So how can this happen? We don't see the lease transfer back to n4 from n24, so it must have been grabbed by n4 when it appeared to be expired. But n24 clearly thinks it's the leaseholder at the time of the assertion failure. This looks like a violation of the lease disjointness invariant.

I think it is. I think n24 must have accidentally shortened its lease when promoting an expiration-based lease to an epoch-based lease. We know that n24 was having a terrible time heartbeating liveness around this time due to a slow disk:
```
drt-large-0015> W240430 18:12:08.280372 769 kv/kvserver/liveness/liveness.go:752 ⋮ [T1,Vsystem,n24,liveness-hb] 1972736 slow heartbeat took 3.00355352s; err=context deadline exceeded
drt-large-0015> W240430 18:12:08.280422 769 kv/kvserver/liveness/liveness.go:668 ⋮ [T1,Vsystem,n24,liveness-hb] 1972737 failed node liveness heartbeat: operation ‹"node liveness heartbeat"› timed out after 3.004s (given timeout 3s): context deadline exceeded
drt-large-0015> W240430 18:12:08.716519 6899959724 kv/kvserver/liveness/liveness.go:752 ⋮ [T1,Vsystem,n24,s369,r45081/25:‹/Table/134/1/"`"/1425{14…-20…}›] 1972769 slow heartbeat took 10.731777395s; err=heartbeat failed on epoch increment
drt-large-0015> W240430 18:12:08.748576 6899959725 kv/kvserver/liveness/liveness.go:752 ⋮ [T1,Vsystem,n24,s369,r44911/21:‹/Table/134/1/"`"/6{196…-200…}›] 1972771 slow heartbeat took 10.763823006s; err=heartbeat failed on epoch increment
drt-large-0015> W240430 18:12:08.802743 6899960152 kv/kvserver/liveness/liveness.go:752 ⋮ [T1,Vsystem,n24,s377,r39579/28:‹/Table/134/1/"`"/882{02…-47…}›] 1972777 slow heartbeat took 10.765866791s; err=heartbeat failed on epoch increment
drt-large-0015> W240430 18:12:08.864379 6899960079 kv/kvserver/liveness/liveness.go:752 ⋮ [T1,Vsystem,n24,s380,r31902/24:‹/Table/129/1/"`"/25{376…-400…}›] 1972794 slow heartbeat took 10.815097374s; err=heartbeat failed on epoch increment
```
This shortening of the lease's effective expiration is a bug, but I don't see any logic preventing it. It could then be possible that n24's slowness prevented it from hearing about the success or failure of the promotion from the expiration-based lease to the epoch-based lease, so it used the longer expiration-based lease to evaluate a closed timestamp update, which then observes the regression.

There's still something that is not quite clicking about how n24 could have proposed a lease request to upgrade from expiration to epoch, not heard the result, and then still gotten past the `ProposalsInFlight` check in `BumpSideTransportClosed`. I can't quite figure that out. However, I don't think that's fatal to this theory, as the overlapping leases are enough of a smoking gun.

----

The fix here will be making sure we don't regress the expiration of a lease when transitioning from an expiration-based lease to an epoch-based lease. I'll try to demonstrate that this is possible. If it is and I haven't missed something, then this will be a longstanding but, and one made more common in v22.2 by 5b0bdc5c.

To address this, we'll want some logic in `pendingLeaseRequest.requestLease` along the following lines:
```go
if isExtension && 
        prevLease.Type == LeaseExpiration && nextLease.Type == LeaseEpoch && 
        nextLease.Liveness.Expiration.Less(prevLease.Expiration()) {
    NodeLiveness.Heartbeat(nextLease.Liveness)
}
```

That [parallels this existing logic](https://github.com/cockroachdb/cockroach/blob/d2dca05d14624ccadc4f23224648e239322e14af/pkg/kv/kvserver/replica_range_lease.go#L517).

**Comment by nvanbenschoten** [comment]
I've confirmed that KVNemesis can trigger the following assertion:
```go
diff --git a/pkg/kv/kvserver/replica_range_lease.go b/pkg/kv/kvserver/replica_range_lease.go
index e4c3b3f3a95..37bf4a7d184 100644
--- a/pkg/kv/kvserver/replica_range_lease.go
+++ b/pkg/kv/kvserver/replica_range_lease.go
@@ -356,6 +356,9 @@ func (p *pendingLeaseRequest) InitOrJoinRequest(
                        return llHandle
                }
                reqLease.Epoch = l.Epoch
+               if extension && status.Lease.Type() == roachpb.LeaseExpiration && l.Expiration.ToTimestamp().Less(status.Expiration()) {
+                       log.Fatalf(ctx, "preventing possible lease expiration regression; cur=%+v, next=%+v, liveness=%+v", status.Lease, reqLease, l)
+               }
        }

        var leaseReq kvpb.Request
```
